### PR TITLE
[TEP-0100] Add new `updatePipelineRunStatusFromChildRefs` function

### PR DESCRIPTION
**Pull request by abayer**
# Changes

And of course, this is part of https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md,
building on a pile of other PRs (#4705, #4734, #4753, #4757).

This adds a new function to `pkg/reconciler/pipelinerun/pipelinerun.go`,
specifically for updating `pr.Status.ChildReferences` during reconciliation. It's
analogous to the existing `updatePipelineRunStatusFromTaskRuns` and
`updatePipelineRunStatusFromRuns` functions. This PR doesn't actually call the
new function - behavior is exactly the same. But it adds the new function,
along with other functions it depends on. In the final step of the
implementation, these other functions will also be used in `...FromTaskRuns`
and/or `...FromRuns`.

I also reworked `pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go`
to improve its test fixtures, so that they're easier to reuse and instantiated
via YAML parsing as much as possible.

/kind tep
/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go [hunk 1]
```diff
@@ -422,6 +422,21 @@ type ChildStatusReference struct {
 	WhenExpressions []WhenExpression `json:"whenExpressions,omitempty"`
 }
 
+// GetConditionChecksAsMap returns a map representation of this ChildStatusReference's ConditionChecks, in the same form
+// as PipelineRunTaskRunStatus.ConditionChecks.
+func (cr ChildStatusReference) GetConditionChecksAsMap() map[string]*PipelineRunConditionCheckStatus {
```
Comment (lbernick) [inline pkg/apis/pipeline/v1beta1/pipelinerun_types.go:427]: nit: this function can be named "GetConditionChecks" ("asMap" is redundant)

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 2]
```diff
@@ -1196,6 +1196,55 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 	return nil
 }
 
+// filterTaskRunsForPipelineRun takes a slice of TaskRuns with the given PipelineRun marked as owner via the "tekton.dev/pipelineRun"
+// label, and returns a slice of TaskRuns directly specified in the PipelineRun and a map of condition TaskRuns keyed by
+// the PipelineTask name they relate to.
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun.go:1201]: This docstring could be rewritten as "filterTaskRunsForPipelineRun returns TaskRuns owned by the PipelineRun and their condition checks"

**Comment by abayer** [comment]
/retest

GitHub's API has been flaking out sporadically lately...

File: pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go [hunk 3]
```diff
@@ -41,168 +44,119 @@ type updateStatusTaskRunsData struct {
...
-	prccs2Recovered := map[string]*v1beta1.PipelineRunConditionCheckStatus{
-		"pr-task-2-running-condition-check-xxyyy": {
-			ConditionName: "running-condition-0",
-		},
-	}
-	prccs3Recovered := map[string]*v1beta1.PipelineRunConditionCheckStatus{
-		"pr-task-3-successful-condition-check-xxyyy": {
-			ConditionName: "successful-condition-0",
-		},
+func getUpdateStatusTaskRunsData(t *testing.T) updateStatusTaskRunsData {
```
Comment (lbernick) [inline pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go:47]: cleanup like swapping over to yaml parsing should be in a separate PR, although OK if you don't want to do it this time since the original PR has already been split several times

**Review by lbernick — COMMENTED** [review]
/lgtm

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go (hunk 1 above)
Comment (abayer) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go:427]: 👍

File: pkg/reconciler/pipelinerun/pipelinerun.go (hunk 2 above)
Comment (abayer) [reply pkg/reconciler/pipelinerun/pipelinerun.go:1201]: 👍

File: pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go (hunk 3 above)
Comment (abayer) [reply pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go:47]: I thought it made sense to just do it here. =)

Comment (abayer) [reply pkg/reconciler/pipelinerun/pipelinerun_updatestatus_test.go:47]: (also I originally did the ->YAML parsing change before I split this and the remainder of the implementation up, so it was markedly simpler to just bring it along)

**Review by vdemeester — APPROVED** [review]

**Review by lbernick — COMMENTED** [review]
/lgtm
